File: modules/knowledge-graph-generator.js
const path = require('path');
const fs = require('fs-extra');

/**
 * Knowledge Graph Generator
 * Creates a graph representation of system entities and their relationships
 */
class KnowledgeGraphGenerator {
  constructor(metrics, options = {}) {
    this.metrics = metrics;
    this.options = Object.assign({
      outputDir: 'knowledge_base',
      graphFormat: 'json', // 'json', 'cytoscape', 'neo4j'
    }, options);
    
    // Initialize graph data structure
    this.graph = {
      nodes: [],
      edges: []
    };
    
    // Track node IDs to prevent duplicates
    this.nodeIds = new Set();
  }
  
  /**
   * Generate a knowledge graph from system data
   */
  async generateKnowledgeGraph(baseDir, sourceSystems) {
    console.log("Generating knowledge graph...");
    
    const outputDir = path.join(baseDir, this.options.outputDir);
    await fs.ensureDir(outputDir);
    
    // Clear existing graph
    this.graph = { nodes: [], edges: [] };
    this.nodeIds.clear();
    
    // Process models and their relationships
    await this.processModelsAndRelationships();
    
    // Process file dependencies
    await this.processFileDependencies();
    
    // Process architecture components
    await this.processArchitectureComponents();
    
    // Generate cross-system integration nodes
    await this.generateIntegrationNodes();
    
    // Write graph to file
    const graphPath = path.join(outputDir, 'knowledge_graph.json');
    await fs.writeFile(graphPath, JSON.stringify(this.graph, null, 2));
    
    // Generate visualization if needed
    if (this.options.generateVisualization) {
      await this.generateVisualization(outputDir);
    }
    
    console.log(`Knowledge graph generated with ${this.graph.nodes.length} nodes and ${this.graph.edges.length} edges`);
    
    return graphPath;
  }
  
  /**
   * Process models and their relationships
   */
  async processModelsAndRelationships() {
    console.log("Processing models and relationships...");
    
    // Process models from both systems
    for (const systemName of Object.keys(this.metrics.sourceSystems || {})) {
      const models = this.metrics.sourceSystems[systemName]?.models?.details || [];
      
      models.forEach(model => {
        this.addNode({
          id: `${systemName}.model.${model.name}`,
          label: model.name,
          type: 'model',
          system: systemName,
          attributes: {
            ...model,
            system: systemName
          }
        });
      });
    }
    
    // Process relationships
    const relationships = this.metrics.relationships || [];
    relationships.forEach(rel => {
      let fromSystem = 'unknown';
      let toSystem = 'unknown';
      
      // Determine which system each model belongs to
      for (const systemName of Object.keys(this.metrics.sourceSystems || {})) {
        const models = this.metrics.sourceSystems[systemName]?.models?.details || [];
        
        if (models.some(m => m.name === rel.from)) {
          fromSystem = systemName;
        }
        
        if (models.some(m => m.name === rel.to)) {
          toSystem = systemName;
        }
      }
      
      // Add edge between models
      this.addEdge({
        source: `${fromSystem}.model.${rel.from}`,
        target: `${toSystem}.model.${rel.to}`,
        label: rel.type,
        type: 'relationship',
        attributes: {
          relationshipType: rel.type,
          notes: rel.notes || ''
        }
      });
    });
  }
  
  /**
   * Process file dependencies
   */
  async processFileDependencies() {
    console.log("Processing file dependencies...");
    
    // We'll use controllers and their dependencies as an example
    for (const systemName of Object.keys(this.metrics.sourceSystems || {})) {
      const controllers = this.metrics.sourceSystems[systemName]?.controllers?.details || [];
      
      controllers.forEach(controller => {
        const controllerId = `${systemName}.controller.${controller.name}`;
        
        this.addNode({
          id: controllerId,
          label: controller.name,
          type: 'controller',
          system: systemName,
          attributes: {
            ...controller,
            system: systemName
          }
        });
        
        // Connect controllers to models they use
        const usedModels = controller.usedModels || [];
        usedModels.forEach(model => {
          // Try to find the model node
          const modelId = `${systemName}.model.${model}`;
          
          // Add edge only if the model node exists
          if (this.nodeIds.has(modelId)) {
            this.addEdge({
              source: controllerId,
              target: modelId,
              label: 'uses',
              type: 'dependency',
              attributes: {
                dependencyType: 'controller_model'
              }
            });
          }
        });
      });
    }
  }
  
  /**
   * Process architecture components
   */
  async processArchitectureComponents() {
    console.log("Processing architecture components...");
    
    // Define core architectural components for each system
    const architectureComponents = {
      canvas: [
        { id: 'auth', label: 'Authentication', type: 'component' },
        { id: 'courses', label: 'Courses', type: 'component' },
        { id: 'assignments', label: 'Assignments', type: 'component' },
        { id: 'gradebook', label: 'Gradebook', type: 'component' },
        { id: 'modules', label: 'Modules', type: 'component' },
        { id: 'files', label: 'Files', type: 'component' },
        { id: 'api', label: 'API', type: 'interface' }
      ],
      discourse: [
        { id: 'auth', label: 'Authentication', type: 'component' },
        { id: 'topics', label: 'Topics', type: 'component' },
        { id: 'posts', label: 'Posts', type: 'component' },
        { id: 'categories', label: 'Categories', type: 'component' },
        { id: 'users', label: 'Users', type: 'component' },
        { id: 'plugins', label: 'Plugins', type: 'component' },
        { id: 'api', label: 'API', type: 'interface' }
      ]
    };
    
    // Add architectural components as nodes
    for (const [systemName, components] of Object.entries(architectureComponents)) {
      components.forEach(component => {
        this.addNode({
          id: `${systemName}.arch.${component.id}`,
          label: component.label,
          type: 'architecture',
          system: systemName,
          attributes: {
            componentType: component.type,
            system: systemName
          }
        });
      });
      
      // Add relationships between components
      if (systemName === 'canvas') {
        // Example Canvas component relationships
        this.addEdge({
          source: `${systemName}.arch.auth`,
          target: `${systemName}.arch.courses`,
          label: 'authenticates',
          type: 'arch_relation'
        });
        
        this.addEdge({
          source: `${systemName}.arch.courses`,
          target: `${systemName}.arch.assignments`,
          label: 'contains',
          type: 'arch_relation'
        });
        
        this.addEdge({
          source: `${systemName}.arch.assignments`,
          target: `${systemName}.arch.gradebook`,
          label: 'feeds',
          type: 'arch_relation'
        });
        
        this.addEdge({
          source: `${systemName}.arch.api`,
          target: `${systemName}.arch.courses`,
          label: 'exposes',
          type: 'arch_relation'
        });
      } else if (systemName === 'discourse') {
        // Example Discourse component relationships
        this.addEdge({
          source: `${systemName}.arch.auth`,
          target: `${systemName}.arch.users`,
          label: 'authenticates',
          type: 'arch_relation'
        });
        
        this.addEdge({
          source: `${systemName}.arch.categories`,
          target: `${systemName}.arch.topics`,
          label: 'contains',
          type: 'arch_relation'
        });
        
        this.addEdge({
          source: `${systemName}.arch.topics`,
          target: `${systemName}.arch.posts`,
          label: 'contains',
          type: 'arch_relation'
        });
        
        this.addEdge({
          source: `${systemName}.arch.api`,
          target: `${systemName}.arch.topics`,
          label: 'exposes',
          type: 'arch_relation'
        });
      }
      
      // Connect architecture components to related models
      this.connectArchitectureToModels(systemName);
    }
  }
  
  /**
   * Connect architecture components to related models
   */
  connectArchitectureToModels(systemName) {
    const models = this.metrics.sourceSystems[systemName]?.models?.details || [];
    
    // Map models to architecture components based on naming patterns
    models.forEach(model => {
      const modelId = `${systemName}.model.${model.name}`;
      let connectedToArch = false;
      
      // Define mapping rules based on system
      const mappingRules = systemName === 'canvas' ? {
        'auth': /user|account|profile|session|authent/i,
        'courses': /course|section|enrollment/i,
        'assignments': /assign|submiss|quiz|rubric/i,
        'gradebook': /grade|score|rubric/i,
        'modules': /module|content|page|item/i,
        'files': /file|folder|attachment/i
      } : {
        'auth': /user|account|profile|session|authent/i,
        'topics': /topic|thread/i,
        'posts': /post|reply|message/i,
        'categories': /categor|tag|group/i,
        'users': /user|profile|account/i,
        'plugins': /plugin|extension|add_on/i
      };
      
      // Connect model to matching architecture component
      Object.entries(mappingRules).forEach(([component, pattern]) => {
        if (pattern.test(model.name)) {
          const componentId = `${systemName}.arch.${component}`;
          
          if (this.nodeIds.has(componentId)) {
            this.addEdge({
              source: modelId,
              target: componentId,
              label: 'implements',
              type: 'architecture_mapping',
              attributes: {
                confidence: pattern.test(model.name.toLowerCase()) ? 'high' : 'medium',
                mapping_type: 'semantic'
              }
            });
            connectedToArch = true;
          }
        }
      });
    });
  }
  
  /**
   * Add a node to the graph
   */
  addNode(node) {
    if (!this.nodeIds.has(node.id)) {
      this.graph.nodes.push(node);
      this.nodeIds.add(node.id);
    }
  }
  
  /**
   * Add an edge to the graph
   */
  addEdge(edge) {
    this.graph.edges.push(edge);
  }
  
  /**
   * Generate cross-system integration nodes
   */
  async generateIntegrationNodes() {
    console.log("Generating integration nodes...");
    
    // Example: Add integration nodes between Canvas and Discourse
    this.addNode({
      id: 'integration.canvas_discourse',
      label: 'Canvas-Discourse Integration',
      type: 'integration',
      attributes: {
        description: 'Integration between Canvas and Discourse systems'
      }
    });
    
    // Example: Add edges representing integration points
    this.addEdge({
      source: 'canvas.arch.api',
      target: 'integration.canvas_discourse',
      label: 'integrates_with',
      type: 'integration_point'
    });
    
    this.addEdge({
      source: 'discourse.arch.api',
      target: 'integration.canvas_discourse',
      label: 'integrates_with',
      type: 'integration_point'
    });
  }
  
  /**
   * Generate visualization of the knowledge graph
   */
  async generateVisualization(outputDir) {
    console.log("Generating visualization...");
    
    // Example: Generate a simple HTML visualization
    const htmlContent = `
      <!DOCTYPE html>
      <html>
      <head>
        <title>Knowledge Graph Visualization</title>
        <style>
          body { font-family: Arial, sans-serif; }
          #graph { width: 100%; height: 100vh; }
        </style>
      </head>
      <body>
        <h1>Knowledge Graph Visualization</h1>
        <div id="graph"></div>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/cytoscape/3.19.1/cytoscape.min.js"></script>
        <script>
          const graphData = ${JSON.stringify(this.graph, null, 2)};
          const cy = cytoscape({
            container: document.getElementById('graph'),
            elements: graphData,
            style: [
              { selector: 'node', style: { 'label': 'data(label)' } },
              { selector: 'edge', style: { 'label': 'data(label)', 'curve-style': 'bezier', 'target-arrow-shape': 'triangle' } }
            ],
            layout: { name: 'cose' }
          });
        </script>
      </body>
      </html>
    `;
    
    const htmlPath = path.join(outputDir, 'visualization.html');
    await fs.writeFile(htmlPath, htmlContent);
    
    console.log(`Visualization generated at ${htmlPath}`);
  }
}

module.exports = KnowledgeGraphGenerator;

File: src/modules/knowledge_graph_generator.rs
use std::path::{Path, PathBuf};
use std::collections::{HashMap, HashSet};
use std::fs;
use anyhow::{Result, Context};
use serde::{Serialize, Deserialize};
use log::info;

/// Node in the knowledge graph
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphNode {
    pub id: String,
    pub label: String,
    pub node_type: String,
    pub properties: HashMap<String, String>,
    pub system: Option<String>,
}

/// Edge in the knowledge graph
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphEdge {
    pub id: String,
    pub source: String,
    pub target: String,
    pub label: String,
    pub edge_type: String,
    pub properties: HashMap<String, String>,
}

/// Complete knowledge graph structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeGraph {
    pub nodes: Vec<GraphNode>,
    pub edges: Vec<GraphEdge>,
}

/// Options for knowledge graph generation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeGraphOptions {
    pub output_dir: String,
    pub graph_format: GraphFormat,
}

/// Format for the output graph
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub enum GraphFormat {
    Json,
    Cytoscape,
    Neo4j,
}

impl Default for KnowledgeGraphOptions {
    fn default() -> Self {
        Self {
            output_dir: "knowledge_base".to_string(),
            graph_format: GraphFormat::Json,
        }
    }
}

/// Creates a graph representation of system entities and their relationships
pub struct KnowledgeGraphGenerator<M> {
    metrics: M,
    options: KnowledgeGraphOptions,
    graph: KnowledgeGraph,
    node_ids: HashSet<String>,
}

impl<M> KnowledgeGraphGenerator<M> {
    /// Create a new knowledge graph generator with the given metrics and options
    pub fn new(metrics: M, options: Option<KnowledgeGraphOptions>) -> Self {
        let options = options.unwrap_or_default();
        
        Self {
            metrics,
            options,
            graph: KnowledgeGraph {
                nodes: Vec::new(),
                edges: Vec::new(),
            },
            node_ids: HashSet::new(),
        }
    }
    
    /// Generate a knowledge graph from system data
    pub async fn generate_knowledge_graph(
        &mut self,
        base_dir: &Path,
        source_systems: &HashMap<String, PathBuf>,
    ) -> Result<&KnowledgeGraph> {
        info!("Generating knowledge graph...");
        
        let output_dir = base_dir.join(&self.options.output_dir);
        fs::create_dir_all(&output_dir).context("Failed to create output directory")?;
        
        // Clear existing graph
        self.graph.nodes.clear();
        self.graph.edges.clear();
        self.node_ids.clear();
        
        // Process models and their relationships
        self.process_models_and_relationships()
            .context("Failed to process models and relationships")?;
        
        // Process file dependencies
        self.process_file_dependencies()
            .context("Failed to process file dependencies")?;
        
        // Process architecture components
        self.process_architecture_components()
            .context("Failed to process architecture components")?;
        
        // Generate cross-system integration nodes
        self.generate_integration_nodes()
            .context("Failed to generate integration nodes")?;
        
        // Write graph to file
        self.write_graph_to_file(&output_dir)
            .context("Failed to write graph to file")?;
        
        info!("Knowledge graph generated with {} nodes and {} edges", 
             self.graph.nodes.len(), 
             self.graph.edges.len());
             
        Ok(&self.graph)
    }
    
    /// Process models and their relationships
    fn process_models_and_relationships(&mut self) -> Result<()> {
        // This would process models and add them as nodes in the graph
        // with edges representing their relationships
        
        // Since we don't have the actual implementation details, this is a placeholder
        // In a real implementation, you would:
        // 1. Find all models in your system
        // 2. Extract their properties and relationships
        // 3. Create nodes and edges accordingly
        
        info!("Processing models and relationships");
        
        Ok(())
    }
    
    /// Process file dependencies
    fn process_file_dependencies(&mut self) -> Result<()> {
        // This would analyze import/require statements to build a dependency graph
        
        // Since we don't have the actual implementation details, this is a placeholder
        // In a real implementation, you would:
        // 1. Analyze source files to extract import/require statements
        // 2. Create edges representing these dependencies
        
        info!("Processing file dependencies");
        
        Ok(())
    }
    
    /// Process architecture components
    fn process_architecture_components(&mut self) -> Result<()> {
        // This would add higher-level architectural components as nodes
        // and connect them to the files/models they contain
        
        // Since we don't have the actual implementation details, this is a placeholder
        // In a real implementation, you would:
        // 1. Define architectural components (services, controllers, etc.)
        // 2. Add them as nodes
        // 3. Connect them to their contents
        
        info!("Processing architecture components");
        
        Ok(())
    }
    
    /// Generate cross-system integration nodes
    fn generate_integration_nodes(&mut self) -> Result<()> {
        // This would create nodes and edges for cross-system integration points
        
        // Since we don't have the actual implementation details, this is a placeholder
        // In a real implementation, you would:
        // 1. Identify integration points between systems
        // 2. Create nodes for these integration points
        // 3. Connect them to the relevant systems
        
        info!("Generating integration nodes");
        
        Ok(())
    }
    
    /// Write graph to file in the specified format
    fn write_graph_to_file(&self, output_dir: &Path) -> Result<()> {
        let filename = match self.options.graph_format {
            GraphFormat::Json => "knowledge_graph.json",
            GraphFormat::Cytoscape => "knowledge_graph_cytoscape.json",
            GraphFormat::Neo4j => "knowledge_graph_neo4j.json",
        };
        
        let graph_path = output_dir.join(filename);
        
        // Convert the graph to the appropriate format
        let graph_data = match self.options.graph_format {
            GraphFormat::Json => {
                // Simple JSON format
                serde_json::to_string_pretty(&self.graph)
                    .context("Failed to serialize graph to JSON")?
            },
            GraphFormat::Cytoscape => {
                // Convert to Cytoscape.js format
                let cytoscape_format = self.to_cytoscape_format();
                serde_json::to_string_pretty(&cytoscape_format)
                    .context("Failed to serialize graph to Cytoscape format")?
            },
            GraphFormat::Neo4j => {
                // Convert to Neo4j format (Cypher statements)
                self.to_neo4j_format()
            },
        };
        
        // Write to file
        fs::write(&graph_path, graph_data)
            .context(format!("Failed to write graph to file: {:?}", graph_path))?;
        
        info!("Knowledge graph written to {:?}", graph_path);
        
        Ok(())
    }
    
    /// Add a node to the graph if it doesn't already exist
    fn add_node(&mut self, node: GraphNode) -> Result<()> {
        if !self.node_ids.contains(&node.id) {
            self.node_ids.insert(node.id.clone());
            self.graph.nodes.push(node);
        }
        
        Ok(())
    }
    
    /// Add an edge to the graph
    fn add_edge(&mut self, edge: GraphEdge) -> Result<()> {
        // Validate that source and target nodes exist
        if !self.node_ids.contains(&edge.source) {
            return Err(anyhow::anyhow!("Source node {} does not exist", edge.source));
        }
        
        if !self.node_ids.contains(&edge.target) {
            return Err(anyhow::anyhow!("Target node {} does not exist", edge.target));
        }
        
        self.graph.edges.push(edge);
        
        Ok(())
    }
    
    /// Convert the graph to Cytoscape.js format
    fn to_cytoscape_format(&self) -> serde_json::Value {
        let elements = serde_json::json!({
            "nodes": self.graph.nodes.iter().map(|node| {
                serde_json::json!({
                    "data": {
                        "id": node.id,
                        "label": node.label,
                        "type": node.node_type,
                        // Include any other properties
                        "properties": node.properties,
                        "system": node.system
                    }
                })
            }).collect::<Vec<_>>(),
            "edges": self.graph.edges.iter().map(|edge| {
                serde_json::json!({
                    "data": {
                        "id": edge.id,
                        "source": edge.source,
                        "target": edge.target,
                        "label": edge.label,
                        "type": edge.edge_type,
                        // Include any other properties
                        "properties": edge.properties
                    }
                })
            }).collect::<Vec<_>>()
        });
        
        elements
    }
    
    /// Convert the graph to Neo4j format (Cypher statements)
    fn to_neo4j_format(&self) -> String {
        let mut cypher = String::new();
        
        // Generate node creation statements
        for node in &self.graph.nodes {
            let properties = node.properties.iter()
                .map(|(k, v)| format!("{}: \"{}\"", k, v.replace("\"", "\\\"")))
                .collect::<Vec<_>>()
                .join(", ");
            
            let system = if let Some(sys) = &node.system {
                format!(", system: \"{}\"", sys.replace("\"", "\\\""))
            } else {
                String::new()
            };
            
            let statement = format!(
                "CREATE (n:{} {{id: \"{}\", label: \"{}\", type: \"{}\"{}{}});\n",
                node.node_type,
                node.id,
                node.label.replace("\"", "\\\""),
                node.node_type,
                if properties.is_empty() { "" } else { ", " },
                properties,
                system
            );
            
            cypher.push_str(&statement);
        }
        
        // Generate edge creation statements
        for edge in &self.graph.edges {
            let properties = edge.properties.iter()
                .map(|(k, v)| format!("{}: \"{}\"", k, v.replace("\"", "\\\"")))
                .collect::<Vec<_>>()
                .join(", ");
            
            let statement = format!(
                "MATCH (a {{id: \"{}\"}}), (b {{id: \"{}\"}}) CREATE (a)-[:{} {{id: \"{}\", label: \"{}\"{}{}}}]->(b);\n",
                edge.source,
                edge.target,
                edge.edge_type,
                edge.id,
                edge.label.replace("\"", "\\\""),
                if properties.is_empty() { "" } else { ", " },
                properties
            );
            
            cypher.push_str(&statement);
        }
        
        cypher
    }
}
